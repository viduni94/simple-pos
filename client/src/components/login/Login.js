import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import { connect } from "react-redux";
import { loginUser } from "../../actions/authActions";

import Logo from "../../img/logo.png";

class Login extends Component {
  constructor() {
    super();
    this.state = {
      email: "",
      password: "",
      errors: {}
    };
  }

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  //To redirect to dashboard when logged in
  componentDidMount() {
    if (this.props.auth.isAuthenticated) {
      this.props.history.push("/dashboard");
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.auth.isAuthenticated) {
      this.props.history.push("/dashboard");
    }

    if (nextProps.errors) {
      this.setState({ errors: nextProps.errors });
    }
  }

  onSubmit = e => {
    e.preventDefault();

    const userData = {
      email: this.state.email,
      password: this.state.password
    };

    this.props.loginUser(userData);
  };

  render() {
    const { errors } = this.state;

    return (
      <div className="login">
        <div className="dark-overlay login-inner text-light">
          <div className="container">
            <div className="row">
              <div className="col-md-12 text-center">
                <img src={Logo} alt="logo" className="img-responsive" style={{ width: 350, height: 350, position: "relative" }} />
                <div className="login-form">
                  <div className="row">
                    <div className="col-md-6 m-auto">
                      <form noValidate onSubmit={this.onSubmit}>
                        <div className="form-group">
                          <input type="email" placeholder="Email Address" name="email" value={this.state.email} onChange={this.onChange} className={classnames("form-control form-control-lg", { "is-invalid": errors.email })} />
                          {errors.email && <div className="invalid-feedback">{errors.email}</div>}
                        </div>
                        <div className="form-group">
                          <input type="password" placeholder="Password" name="password" value={this.state.password} onChange={this.onChange} className={classnames("form-control form-control-lg", { "is-invalid": errors.password })} />
                          {errors.password && <div className="invalid-feedback">{errors.password}</div>}
                        </div>
                        <button type="submit" className="btn btn-login btn-block mt-4">
                          Login
                        </button>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  loginUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { loginUser }
)(Login);
