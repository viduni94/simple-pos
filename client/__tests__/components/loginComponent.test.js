import React from "react";
import { shallow } from "enzyme";
import { Login } from "../../src/components/login/Login";

const props = {
  loginUser: jest.fn().mockReturnThis(),
  setActivePage: jest.fn().mockReturnThis(),
  auth: {
    isAuthenticated: ""
  },
  errors: {},
  page: {}
};

describe("Login Component", () => {
  test("renders", done => {
    const wrapper = shallow(<Login {...props} />);
    expect(wrapper).toHaveLength(1);
    done();
  });
  
  const testProps = { ...props, auth: { isAuthenticated: true }, history: [] };
  test("redirects when isAuthenticated is true", done => {
    const wrapper = shallow(<Login {...testProps} />);
    expect(wrapper.instance().props.history.includes("/dashboard")).toBe(true);
    done();
  });

  const testProps1 = { ...props, auth: { isAuthenticated: false }, history: [] };
  test("redirects when isAuthenticated changes to true", done => {
    const wrapper = shallow(<Login {...testProps1} />);
    expect(wrapper.instance().props.history.includes("/dashboard")).toBe(false);
    wrapper.setProps({ auth: { isAuthenticated: true } });
    expect(wrapper.instance().props.history.includes("/dashboard")).toBe(true);
    done();
  });

  test("receive errors", done => {
    const wrapper = shallow(<Login {...props} />);
    wrapper.setProps({ errors: { email: "invalid", password: 'invalid' } });
    expect(wrapper.instance().props.errors.email).toEqual('invalid');
    expect(wrapper.instance().props.errors.password).toEqual('invalid');
    done();
  });

  test("should render initial layout", () => {
    // when
    const component = shallow(<Login {...props} />);
    // then
    expect(component.getElements()).toMatchSnapshot();
  });

  test("when the form is submitted the event is cancelled", done => {
    const wrapper = shallow(<Login {...props} />);
    let prevented = false;

    wrapper.find("form").simulate("submit", {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(prevented).toBe(true);
    done();
  });

  test("should create an entry in component state with the event value", () => {
    // given
    const component = shallow(<Login {...props} />);
    const form = component.find("input").at(1);
    // when
    form.props().onChange({
      target: {
        name: "email",
        value: "myValue"
      }
    });
    // then
    expect(component.state("email")).toEqual("myValue");
  });
});
