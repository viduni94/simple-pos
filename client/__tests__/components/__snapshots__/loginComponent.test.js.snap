// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should render initial layout 1`] = `
Array [
  <div
    className="login"
  >
    <div
      className="dark-overlay login-inner text-light"
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="col-md-12 text-center"
          >
            <img
              alt="logo"
              className="img-responsive"
              src="test-file-stub"
              style={
                Object {
                  "height": 350,
                  "position": "relative",
                  "width": 350,
                }
              }
            />
            <div
              className="login-form"
            >
              <div
                className="row"
              >
                <div
                  className="col-md-6 m-auto"
                >
                  <form
                    noValidate={true}
                    onSubmit={[Function]}
                  >
                    <div
                      className="form-group"
                    >
                      <input
                        className="form-control form-control-lg"
                        name="email"
                        onChange={[Function]}
                        placeholder="Email Address"
                        type="email"
                        value=""
                      />
                    </div>
                    <div
                      className="form-group"
                    >
                      <input
                        className="form-control form-control-lg"
                        name="password"
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
                      />
                    </div>
                    <button
                      className="btn btn-login btn-block mt-4"
                      type="submit"
                    >
                      Login
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
